File: src/main/java/ski/mashiro/command/AdminCommand.java
package ski.mashiro.command;

import net.mamoe.mirai.console.command.CommandSender;
import net.mamoe.mirai.console.command.FriendCommandSender;
import net.mamoe.mirai.console.command.SystemCommandSender;
import net.mamoe.mirai.console.command.java.JCompositeCommand;
import org.jetbrains.annotations.NotNull;
import ski.mashiro.ScheduleQQAdvice;
import ski.mashiro.file.ConfigFile;
import ski.mashiro.file.WhitelistFile;

/**
 * @author MashiroT
 */
public class AdminCommand extends JCompositeCommand {
    public AdminCommand() {
        super(ScheduleQQAdvice.INSTANCE, "admin");
        super.setPrefixOptional(true);
    }

    @SubCommand("add")
    @Description("增加白名单")
    public void addWhitelist(CommandSender sender, String qq) {
        if (hasNoPermission(sender)) {
            return;
        }
        WhitelistFile.WHITELIST.add(Long.valueOf(qq));
        sender.sendMessage("添加成功");
    }

    @SubCommand("del")
    @Description("删除白名单")
    public void delWhitelist(CommandSender sender, String qq) {
        if (hasNoPermission(sender)) {
            return;
        }
        for (int i = 0; i < WhitelistFile.WHITELIST.size(); i++) {
            if (WhitelistFile.WHITELIST.get(i).equals(Long.valueOf(qq))) {
                WhitelistFile.WHITELIST.remove(i);
                sender.sendMessage("删除成功");
                return;
            }
        }
        sender.sendMessage("不在名单内");
    }

    @SubCommand("setLeadTime")
    @Description("设置提前提醒时间")
    public void setLeadTime(CommandSender sender, int leadTime) {
        if (hasNoPermission(sender)) {
            return;
        }
        if (leadTime < 0) {
            sender.sendMessage("leadTime >= 0");
            return;
        }
        ConfigFile.CONFIG.setLeadTime(leadTime);
        sender.sendMessage("修改成功, 插件重新加载后生效");
    }

    @SubCommand("setUrl")
    @Description("设置接口地址")
    public void setUrl(CommandSender sender, String url) {
        if (hasNoPermission(sender)) {
            return;
        }
        ConfigFile.CONFIG.setUrl(url);
        sender.sendMessage("修改成功, 请确保地址格式正确, 插件重新加载后生效");
    }

    @SubCommand("reload")
    public void reload(@NotNull CommandSender sender) {
        if (hasNoPermission(sender)) {
            return;
        }
        ConfigFile.loadConfig();
        sender.sendMessage("重载成功");
    }

    private boolean hasNoPermission(CommandSender sender) {
        if (!(sender instanceof SystemCommandSender)) {
            if (!(sender instanceof FriendCommandSender)) {
                return true;
            }
        }
        return ConfigFile.CONFIG.getOwner() != sender.getUser().getId();
    }
}

File: src/main/kotlin/ski/mashiro/command/AdminCommand.kt
package ski.mashiro.command

import net.mamoe.mirai.console.command.CommandSender
import net.mamoe.mirai.console.command.CompositeCommand
import net.mamoe.mirai.console.command.descriptor.ExperimentalCommandDescriptors
import net.mamoe.mirai.console.command.isNotConsole
import net.mamoe.mirai.console.command.isUser
import ski.mashiro.CourseScheduleQQAdvice
import ski.mashiro.data.CourseData
import ski.mashiro.data.UserData
import ski.mashiro.file.ConfigOp
import java.time.LocalDateTime

object AdminCommand : CompositeCommand(
    CourseScheduleQQAdvice, "admin"
) {
    @ExperimentalCommandDescriptors
    override val prefixOptional: Boolean
        get() = true

    private val logger = CourseScheduleQQAdvice.logger

    @SubCommand("addWhitelist")
    suspend fun addWhitelist(sender: CommandSender, qq: Long) {
        if (hasNoPerm(sender)) {
            logger.warning("${sender.name} 尝试越权操作")
            return
        }
        if (ConfigOp.config.whitelist.contains(qq)) {
            sender.sendMessage("该用户已在白名单内")
            return
        }
        ConfigOp.config.whitelist.add(qq)
        sender.sendMessage("添加成功")
    }

    @SubCommand("delWhitelist")
    suspend fun delWhitelist(sender: CommandSender, qq: Long) {
        if (hasNoPerm(sender)) {
            logger.warning("${sender.name} 尝试越权操作")
            return
        }
        if (ConfigOp.config.whitelist.contains(qq)) {
            sender.sendMessage("该用户不在白名单内")
            return
        }
        ConfigOp.config.whitelist.remove(qq)
        sender.sendMessage("移除成功")
    }

    @SubCommand("setOwner")
    suspend fun setOwner(sender: CommandSender, qq: Long) {
        if (hasNoPerm(sender)) {
            logger.warning("${sender.name} 尝试越权操作")
            return
        }
        ConfigOp.config.ownerQQ = qq
        sender.sendMessage("更改Owner成功")
    }

    @SubCommand("setApiAddress")
    suspend fun setApiAddress(sender: CommandSender, apiAddress: String) {
        if (hasNoPerm(sender)) {
            logger.warning("${sender.name} 尝试越权操作")
            return
        }
        ConfigOp.config.apiAddress = apiAddress
        sender.sendMessage("更改apiAddress成功")
    }

    @SubCommand("refresh", "ref")
    suspend fun refreshCourseData(sender: CommandSender) {
        if (hasNoPerm(sender)) {
            logger.warning("${sender.name} 尝试越权操作")
            return
        }
        val now = LocalDateTime.now()
        UserData.userMap.values.forEach { user ->
            CourseData.getCourseData(now, user)
        }
        sender.sendMessage("刷新成功")
    }

    private fun hasNoPerm(sender: CommandSender): Boolean {
        return sender.isNotConsole() && sender.isUser() && ConfigOp.config.ownerQQ != sender.user.id
    }
}